### DISCUSSION TITLE
Duplicate key and put_many

### DISCUSSION THREAD

**Discussion by aklaver** [deta/cloud-docs#292] [Help] [answered]
Using Python SDK, deta==1.1.0:
```
from deta import Deta
deta = Deta("<some_key>")
db_put = deta.Base("put_db")
put_data = [{"key": '1', "fld_1": "dog"}, {"key": '2', "fld_1": "cat"}, {"key": '1', "fld_1": "fish"}]
db_put.put_many(put_data)
HTTPError: HTTP Error 400: Bad Request

put_data = [{"key": '1', "fld_1": "dog"}, {"key": '2', "fld_1": "cat"}, {"key": '3', "fld_1": "fish"}]
db_put.put_many(put_data)
{'processed': {'items': [{'fld_1': 'dog', 'key': '1'},
   {'fld_1': 'cat', 'key': '2'},
   {'fld_1': 'fish', 'key': '3'}]}}
```
From here [Put](https://docs.deta.sh/docs/base/sdk#put):

>If an item already exists under a given key, put will replace this item.

I assumed a duplicate key in `put_many` would not be an issue. Obviously I was wrong. Is this expected behavior or a bug?

**Comment by lemonyte** [comment, marked as answer]
I believe the HTTP API only supports duplicate keys in _separate_ requests. If you `put` two items with the same key in one request, how does the API know which one to keep and which one to overwrite?

**Comment by aklaver** [comment]
The same way separate `put`'s does by order of entry.  1) `deta` has no way of knowing in either case whether the order is correct. That is on the user. 2) If the item already exists the first entry would be an overwrite also. 3) `put` as described is a way to an `UPSERT`. If you want a duplicate error you use `insert`.

**Reply by aavshr** [reply]
hi @aklaver this is intended behavior as a `put_many` is taken as a single write, not as a set ordered writes. The order of the items in the array is not explicit enough to be taken as intentional for the API.

**Reply by aklaver** [reply]
That does not make sense. The API has no way of knowing whether  individual `put`'s are in the intended order either. The same process the creates the data for input can do that out of order before either iterating over the data with `put` or batching the data with `put_many`.  The POLA would seem to me that doing a `put` would behave the same in each case.  And as I mentioned earlier if double entry is something to be avoided then there is `insert`.  If this is not going to be changed then the documentation needs to be updated to flag the behavior change.
